sdk/search/Azure.Search.Documents/src/Utilities/AggregateDictionary.cs

```diff
@@ -0,0 +1,311 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Azure.Search.Documents
+{
+    /// <summary>
+    /// Aggregates multiple dictionaries as a single dictionary.
+    /// </summary>
+    /// <typeparam name="TKey">The type of key.</typeparam>
+    /// <typeparam name="TValue">The type of value.</typeparam>
+    internal abstract class AggregateDictionary<TKey, TValue> : IDictionary<TKey, TValue>
+    {
+        private readonly IEnumerable<IDictionary<TKey, TValue>> _dictionaries;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="AggregateDictionary{TKey, TValue}"/> class.
+        /// </summary>
+        /// <param name="dictionaries">The dictionaries to aggregate as a single dictionary.</param>
+        /// <exception cref="ArgumentNullException"><paramref name="dictionaries"/> is null.</exception>
+        public AggregateDictionary(IEnumerable<IDictionary<TKey, TValue>> dictionaries)
+        {
+            _dictionaries = dictionaries ?? throw new ArgumentNullException(nameof(dictionaries));
+
+            // Assume only 2 elements.
+            const int capacity = 2;
+
+            List<ICollection<TKey>> keys = new List<ICollection<TKey>>(capacity);
+            List<ICollection<TValue>> values = new List<ICollection<TValue>>(capacity);
+
+            foreach (IDictionary<TKey, TValue> dictionary in _dictionaries)
+            {
+                keys.Add(dictionary.Keys);
+                values.Add(dictionary.Values);
+            }
+
+            Keys = new ReadOnlyAggregateCollection<TKey>(keys);
+            Values = new ReadOnlyAggregateCollection<TValue>(values);
```

Comment:
`ToList` makes a snapshot. `Keys` and `Values` has to grow and shrink. If someone gets these references and holds onto them, a snapshot will not suffice.

The aggregate already does a sufficient job to avoid duplicates. It may not be completely impossible, but given that the JSON source object can't have duplicates (invalid JSON), the two source dictionaries we use shouldn't either. And if customers try to add a duplicate, the aggregate detects this and reacts accordingly. Anything else would be exceptional, thus requiring debugging and appropriate resolution.

Reply:
OK, I see `KeysAndValuesGrow` is testing that.